### PR TITLE
Refresh docker backend

### PR DESCRIPTION
**Pull request by meatballhat**
to hopefully:

- work with current stable docker
- work correctly when run on separate host

### CODE REVIEW

File: backend/docker.go [hunk 1]
```diff
@@ -312,36 +323,121 @@ func (i *dockerInstance) sshClient() (*ssh.Client, error) {
...
+}
+
+func (i *dockerInstance) runScriptExec(ctx gocontext.Context, output io.Writer) (*RunResult, error) {
+	createExecOpts := docker.CreateExecOptions{
+		// TODO: remove hardcoded shenanigans
+		AttachStdin:  false,
+		AttachStdout: true,
+		AttachStderr: true,
+		Tty:          true,
+		Cmd:          []string{"bash", "-l", "-c", "cd /home/travis && exec bash build.sh"},
```
Comment (meatballhat) [inline backend/docker.go]: I'm pretty sure cd'ing to `/home/travis` is unnecessary... 

pretty sure...

File: backend/docker.go [hunk 2]
```diff
@@ -312,36 +323,121 @@ func (i *dockerInstance) sshClient() (*ssh.Client, error) {
...
 	}
 
-	f, err := sftp.Create("build.sh")
+	// fmt.Printf("---> Uploading to container: %#v\n", uploadOpts)
+
+	return i.client.UploadToContainer(i.container.ID, uploadOpts)
+}
+
+func (i *dockerInstance) RunScript(ctx gocontext.Context, output io.Writer) (*RunResult, error) {
+	if os.Getenv("TRAVIS_WORKER_DOCKER_EXEC") == "1" {
```
Comment (meatballhat) [inline backend/docker.go]: Should this be kept as a first-class feature?

Comment (henrikhodne) [inline backend/docker.go]: Depends on whether Enterprise uses it, I think.

File: backend/docker.go (hunk 1 above)
Comment (henrikhodne) [inline backend/docker.go]: I think the build script expects to be run in `/home/travis`, and I think maybe `docker exec` has `/` as the working directory by default? (I guess a `docker exec pwd` or `docker exec ls` can probably tell you whether that is right)

File: backend/docker.go [hunk 3]
```diff
@@ -28,6 +29,8 @@ var (
 		"CMD":             "command (CMD) to run when creating containers (default \"/sbin/init\")",
 		"MEMORY":          "memory to allocate to each container (0 disables allocation, default \"4G\")",
 		"CPUS":            "cpu count to allocate to each container (0 disables allocation, default 2)",
+		"CPU_SET_SIZE":    "size of available cpu set (default detected locally via runtime.NumCPU)",
+		"EXEC":            "run build script via `docker exec` instead of over ssh (default true)",
```
Comment (igorwwwwwwwwwwwwwwwwwwww) [inline backend/docker.go]: shall we default this to false for now, until we've fully tested it? just to make sure it doesn't automatically get enabled by someone using it (e.g. enterprise customers)

Comment (meatballhat) [inline backend/docker.go]: I'm happy to flip this logic to keep ssh as the default.

**Comment by meatballhat** [comment]
Tested in staging with `TRAVIS_WORKER_DOCKER_EXEC="true"`, fwiw: https://staging.travis-ci.org/meatballhat/yolo-octo-adventure/jobs/506700

**Comment by meatballhat** [comment]
Verified on staging AWS with (redacted) config:

``` bash
export TRAVIS_WORKER_BUILD_API_PARANOID="false"
export TRAVIS_WORKER_BUILD_PARANOID="false"
export TRAVIS_WORKER_PAPERTRAIL_REMOTE_PORT="12345"
export TRAVIS_WORKER_POOL_SIZE="4"
export TRAVIS_WORKER_PROVIDER_NAME="docker"
export TRAVIS_WORKER_QUEUE_NAME="builds.docker"
export TRAVIS_WORKER_AMQP_URI="amqp://secret@hostname:12345/FAFAFAF"
export TRAVIS_WORKER_BUILD_API_URI="https://secret@build-staging.travis-ci.org/script"
export TRAVIS_WORKER_DEBUG="1"
export TRAVIS_WORKER_DOCKER_NATIVE="true"
export TRAVIS_WORKER_DOCKER_ENDPOINT="unix:///var/run/docker.sock"
```

**Comment by meatballhat** [comment]
Example successful run: https://staging.travis-ci.org/meatballhat/yolo-octo-adventure/jobs/506883

**Comment by meatballhat** [comment]
Example successful run with ssh-based execution: https://staging.travis-ci.org/meatballhat/yolo-octo-adventure/jobs/506917

**Comment by meatballhat** [comment]
Also, fwiw:

``` bash
$ docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64
```

**Comment by Lyoness** [comment]
Builds have passed :tada:

**Comment by joshk** [comment]
Oh my gosh! The endpoint can be set to the UNIX socket???????

**Comment by solarce** [comment]
🐑

**Pull request merged**